## Exercise 3  
1. 실습 1번

 손글씨 숫자 MNIST 를 가지고, 98% val_accuracy 보다 크게 나오도록,  convolutional layer 는 딱 1개만 사용하고,  MaxPooling 2D도 한 1개만 사용한다. 그리고 epochs는 20으로 셋팅하고,  val_accuracy 가 위의 수치를 만족하면, "Reached 98% val_accuracy so cancelling training!"를 출력하고, 학습을 멈추도록 모델링과 컴파일을 하시오.



In [1]:
import tensorflow as tf



In [2]:
def train_mnist_conv():
    # YOUR CODE STARTS HERE
    class myCallback(tf.keras.callbacks.Callback) :
      def on_epoch_end(self, epoch, logs={}) :
        if logs.get('val_accuracy') >= 0.98 :
          print('\nReached 98% val_accuracy so cancelling training!')
          self.model.stop_training = True
    
    my_callback = myCallback()
    # YOUR CODE ENDS HERE

    mnist = tf.keras.datasets.mnist
    (training_images, training_labels), (test_images, test_labels) = mnist.load_data()
    # YOUR CODE STARTS HERE
    training_images = training_images / 255.0
    test_images = test_images / 255.0

    training_images = training_images.reshape(60000, 28, 28, 1)
    test_images = test_images.reshape(10000, 28, 28, 1)

    # YOUR CODE ENDS HERE

    model = tf.keras.models.Sequential([
            # YOUR CODE STARTS HERE
            tf.keras.layers.Conv2D(64, 3, activation='relu', input_shape=(28,28,1)),
            tf.keras.layers.MaxPooling2D(2, 2),
            tf.keras.layers.Flatten(), 
            tf.keras.layers.Dense(128, 'relu'),
            tf.keras.layers.Dense(10, 'softmax')
            # YOUR CODE ENDS HERE
    ])

    model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])
    # model fitting
    history = model.fit(
        # YOUR CODE STARTS HERE
        training_images, training_labels, epochs=20, validation_data=(test_images, test_labels),
        callbacks = [my_callback]
        # YOUR CODE ENDS HERE
    )
    # model fitting
    return history.epoch, history.history['accuracy'][-1]



In [3]:
_, _ = train_mnist_conv()

11501568/11490434 [==============================] - 0s 0us/step
Epoch 1/20
1875/1875 [==============================] - 73s 38ms/step - loss: 0.1390 - accuracy: 0.9588 - val_loss: 0.0593 - val_accuracy: 0.9793
Epoch 2/20
1875/1875 [==============================] - ETA: 0s - loss: 0.0480 - accuracy: 0.9852
Reached 98% val_accuracy so cancelling training!
1875/1875 [==============================] - 69s 37ms/step - loss: 0.0480 - accuracy: 0.9852 - val_loss: 0.0410 - val_accuracy: 0.9863


In [4]:
60000 / 32

1875.0

2. 실습 2번

 val_accuracy 가 96% 이상이면 학습을 멈추도록 코드를 수정하여 실행하시오